pkg/metrics/sketch_series.go

```diff
@@ -78,6 +79,118 @@ func (sl SketchSeriesList) MarshalJSON() ([]byte, error) {
+// compressed protobuf marshaled gogen.SketchPayload objects. gogen.SketchPayload is not directly marshaled - instead
+// it's contents are marshaled individually, packed with the appropriate protobuf metadata, and compressed in stream.
+// The resulting payloads (when decompressed) are binary equal to the result of marshaling the whole object at once.
+func (sl SketchSeriesList) MarshalSplitCompress() ([]*[]byte, error) {
+	input := bytes.NewBuffer(make([]byte, 0, 1024))
+	output := bytes.NewBuffer(make([]byte, 0, 1024))
+
+	// The Metadata field of gogen.SketchPayload is never written to - so pack an empty metadata as the footer
+	footer := []byte{0x12, 0}
+
+	compressor, e := stream.NewCompressor(input, output, []byte{}, footer, []byte{})
+	if e != nil {
+		return nil, e
+	}
+	payloads := []*[]byte{}
+	precompressionBuf := make([]byte, 1024)
+
+	dsl := make([]gogen.SketchPayload_Sketch_Dogsketch, 1)
+	for _, ss := range sl {
+		if len(ss.Points) > cap(dsl) {
+			dsl = make([]gogen.SketchPayload_Sketch_Dogsketch, len(ss.Points))
+		}
+
+		for i, p := range ss.Points {
+			b := p.Sketch.Basic
+			k, n := p.Sketch.Cols()
+			dsl[i] = gogen.SketchPayload_Sketch_Dogsketch{
+				Ts:  p.Ts,
+				Cnt: b.Cnt,
+				Min: b.Min,
+				Max: b.Max,
+				Avg: b.Avg,
+				Sum: b.Sum,
+				K:   k,
+				N:   n,
+			}
+		}
+
+		sketch := gogen.SketchPayload_Sketch{
+			Metric:      ss.Name,
+			Host:        ss.Host,
+			Tags:        ss.Tags,
+			Dogsketches: dsl[:len(ss.Points)],
+		}
+
+		// Pack the protobuf metadata - see SketchPayload.MarshalTo in agent_payload.pb.go for reference.
+		i := 0
+		// Magic number that occurs before the varint encoding
+		precompressionBuf[i] = 0xa
+		i++
+		i = encodeVarintAgentPayload(precompressionBuf, i, uint64(sketch.Size()))
+
+		// Resize the pre-compression buffer if needed
+		totalItemSize := sketch.Size() + i
+		if totalItemSize > cap(precompressionBuf) {
+			precompressionBuf = append(precompressionBuf, make([]byte, totalItemSize-cap(precompressionBuf))...)
+			precompressionBuf = precompressionBuf[:cap(precompressionBuf)]
+		}
+
+		// Marshal the sketch to the precompression buffer after the metadata
+		_, e := sketch.MarshalTo(precompressionBuf[i:])
+		if e != nil {
+			return nil, e
+		}
+
+		// Compress the protobuf metadata and the marshaled sketch
+		e = compressor.AddItem(precompressionBuf[:totalItemSize])
+		if e == stream.ErrItemTooBig || e == stream.ErrPayloadFull {
+			// Since the compression buffer is full - flush it and rotate
+			payload, e := compressor.Close()
+			if e != nil {
+				return nil, e
+			}
+			payloads = append(payloads, &payload)
+			input.Reset()
+			output.Reset()
+			compressor, e = stream.NewCompressor(input, output, []byte{}, footer, []byte{})
+			if e != nil {
+				return nil, e
+			}
+
+			// Add it to the new compression buffer - since this is a new compressor it should never overflow.
```

Comment:
if there's a very large item it might overflow, so let's still track it when it does